feat(na): clarify when to use docker vs k8s, rename docker page (#1873)
medic/cht-docs: content/en/hosting/kubernetes-vs-docker.md

title: Kubernetes vs Docker
linkTitle: Kubernetes vs Docker
weight: 5
aliases:
  -  /hosting/vertical-vs-horizontal
---

{{< hextra/hero-subtitle >}}
  Options for installing CHT applications
{{< /hextra/hero-subtitle >}}

{{< callout >}}

> Medic previously suggest Kubernetes for all production deployments - this is no longer the case.  Further, Medic no longer suggests multi-node CouchDB.

Reviewer: ```suggestion
For older versions of the CHT, Medic recommended Kubernetes for all production deployments. This is no longer the recommendation for newer versions of the CHT. Further, multi-node CouchDB is no longer recommended.
```

{{< /callout >}}

> Medic now recommends [Docker](/hosting/4.x/docker/) for most deployments. Use [Kubernetes](/hosting/4.x/kubernetes/) for multi-tenant or specific hardware constrained deployments. For more details on the research behind suggesting Docker and single-node CouchDB, please see [this forum post](https://forum.communityhealthtoolkit.org/t/investigate-adding-more-shards-as-a-potential-avenue-for-improved-performance/4831?u=mrjones).

Reviewer: ```suggestion
Medic recommends [Docker](/hosting/4.x/docker/) for most deployments. Use [Kubernetes](/hosting/4.x/kubernetes/) for multi-tenant or specific hardware constrained deployments. For more details on the research behind suggesting Docker and single-node CouchDB, see [this forum post](https://forum.communityhealthtoolkit.org/t/investigate-adding-more-shards-as-a-potential-avenue-for-improved-performance/4831?u=mrjones).
```

Reviewer: Per the [docs style guide](https://docs.communityhealthtoolkit.org/community/contributing/docs/style-guide/#simple-and-direct-language), we should avoid please.

Author: Thanks for the attention to detail - I'll revisit that style guide as a refresher!

> That said, deployments should use the technology they know best. If they plan for 20 instances but really doen't want to use Kubernetes, that's fine.  Conversely, if an NGO only uses Kubernetes and wants to deploy a single instance of the CHT - that's also fine.

Reviewer: ```suggestion
However, deployments should use the technology they are most familiar and comfortable with. If they plan for 20 CHT instances but don't want to use Kubernetes, that's fine. Conversely, if a hosting organisation uses Kubernetes and intends to deploy a single instance of the CHT, that's also fine.
```

> Application development for both [CHT 3.x]({{< relref "hosting/3.x/app-developer" >}}) and [CHT 4.x]({{< relref "hosting/4.x/app-developer" >}}) should use Docker.

Reviewer: ```suggestion
{{< callout >}}
 Application development for both [CHT 3.x]({{< relref "hosting/3.x/app-developer" >}}) and [CHT 4.x]({{< relref "hosting/4.x/app-developer" >}}) should use Docker.
{{< /callout >}}
```

* Service discovery which  enables to route public requests to deployments within the cluster
* Integration with hypervisors such as VMWare and ESX. This is possible due to CRD support in Kubernetes that allows 3rd parties to create integrations
* Helm support and integration that makes it easy to easily deploy applications on to the cluster
* Highly efficient snapshot backups when used with a storage area network [SAN](https://en.wikipedia.org/wiki/Storage_area_network) or Amazon's [Elastic Block Storage](https://aws.amazon.com/ebs/) (EBS).

The main components of a Kubernetes CHT deployment include one less service than a Compose based deployment as ingress via `nginx` isn't needed as it's handled externally:

* A pod for each CouchDB instance.
* A CHT API pod.
* A CHT HAProxy Healthcheck pod.
* A CHT HAProxy pod.
* Upgrade Service pod.
* CHT-Sentinel pod.

## Example deployments

Below are examples from 2 real world Ministries of Health (MoH) which have been anonymized and Medic hosted instances.  Users are total provisioned users, fewer users may be active in a given month.

> ### MoH Example 1

Reviewer: I appreciate how illustrative the real-world examples are, thank you!

> Below are examples from 2 real world Ministries of Health (MoH) which have been anonymized and Medic hosted instances.  Users are total provisioned users, fewer users may be active in a given month.

Reviewer: ```suggestion
Below are examples from real world large-scale CHT deployments which have been anonymised.  Users are total provisioned users, fewer users may be active in a given month.
```

Reviewer: I don't see any benefit in mentioning whether Medic is hosted or not. The important information here is the scale and technology used.

Reviewer 2: Can we use average monthly active users instead of provisioned users?

Author: oop - missed Diana's request for average monthly active users - unresolving!

Author: moved all instances to 30 day average active user  in 986f4df8

> ### MoH Example 1

Reviewer: ```suggestion
### CHT Deployment Example 1
```

> This MoH hosts 47 production instances in a national data center running [K3s](https://k3s.io/). They run a simplified k3s set up with 47 isolated installations of k3s to easy management complexity.  Thus each CHT instance is bound to the VM's local storage.  If a VM fails, there is no automatic fail-over and a restore from backup is needed.

Reviewer: ```suggestion
This deployment hosts 47 production instances in a data center running [K3s](https://k3s.io/). It runs a simplified k3s set up with 47 isolated installations of k3s to manage complexity easily.  Thus each CHT instance is bound to the VM's local storage.  If a VM fails, there is no automatic failover, and a restore from backup is needed.
```

|                   |                             |
|------------------:|:----------------------------|

> |    CHT Instances: | 47                          |

Reviewer 2: I think the more important breakdown here was by instance, and how many users / resources / repication apdex they have, at least that's what I was envisioning.

Author: Do you mean for all three examples, or for just Medic hosted examples?  Assuming just medic hosted, maybe breaking out two examples deployments, one small and one big?

|          Hosting: | k3s                         |
|            Cores: | 376 (47VMs x 8 Cores/ea)    |
|              RAM: | 752 GB (47VMx 16 GB/ea)     |
|             Disk: | 4,700 GB  (Max 700, Min 21) |
| Multi-Node Couch: | No                          |
|            Users: | 123.3k                      |

> ### MoH Example 2

Reviewer: ```suggestion
### CHT Deployment Example 2
```

> This MoH hosts 1 of production instance in a national data center running [K3s](https://k3s.io/).

Reviewer: ```suggestion
This deployment hosts 1 CHT production instance in a data center running [K3s](https://k3s.io/).
```

|                   |                           |
|------------------:|:--------------------------|
|    CHT Instances: | 1                         |
|          Hosting: | K3s on VMware)            |
|            Cores: | 32 (4 VMs x 8 cores/ea)   |
|              RAM: | 64 GB (4 VMs x 16GB/ea)   |
|             Disk: | 600 GB (4 VMs x 150GB/ea) |
| Multi-Node Couch: | Yes                       |
|            Users: | 18k                       |

> ### Medic

Reviewer: ```suggestion
### CHT Deployment Example 3
```

> Medic hosts 46 production instances on Amazon [Elastic Kubernetes Service](https://docs.aws.amazon.com/eks/latest/userguide/what-is-eks.html) (EKS).

Reviewer: ```suggestion
This deployment hosts 46 production instances on Amazon [Elastic Kubernetes Service](https://docs.aws.amazon.com/eks/latest/userguide/what-is-eks.html) (EKS).
```

* Orchestration of multiple deployments ensuring to allow easy hosting of multi-tenants. For example you may opt to have a user acceptance testing (UAT), staging and production instances all in one cluster.
* Service discovery which  enables to route public requests to deployments within the cluster
* Integration with hypervisors such as VMWare and ESX. This is possible due to CRD support in Kubernetes that allows 3rd parties to create integrations
* Helm support and integration that makes it easy to easily deploy applications on to the cluster
* Highly efficient snapshot backups when used with a storage area network [SAN](https://en.wikipedia.org/wiki/Storage_area_network) or Amazon's [Elastic Block Storage](https://aws.amazon.com/ebs/) (EBS).

The main components of a Kubernetes CHT deployment include one less service than a Compose based deployment as ingress via `nginx` isn't needed as it's handled externally:

* A pod for each CouchDB instance.
* A CHT API pod.
* A CHT HAProxy Healthcheck pod.
* A CHT HAProxy pod.
* Upgrade Service pod.
* CHT-Sentinel pod.

## Example deployments

> Below are examples from real world large-scale CHT deployments which have been anonymised.  Users are active users for the past 30 days.

Reviewer 2: ```suggestion
Below are examples from real world large-scale CHT deployments which have been anonymized. User counts represent the number of active users within a 30 days interval. 
```

* Service discovery which  enables to route public requests to deployments within the cluster
* Integration with hypervisors such as VMWare and ESX. This is possible due to CRD support in Kubernetes that allows 3rd parties to create integrations
* Helm support and integration that makes it easy to easily deploy applications on to the cluster
* Highly efficient snapshot backups when used with a storage area network [SAN](https://en.wikipedia.org/wiki/Storage_area_network) or Amazon's [Elastic Block Storage](https://aws.amazon.com/ebs/) (EBS).

The main components of a Kubernetes CHT deployment include one less service than a Compose based deployment as ingress via `nginx` isn't needed as it's handled externally:

* A pod for each CouchDB instance.
* A CHT API pod.
* A CHT HAProxy Healthcheck pod.
* A CHT HAProxy pod.
* Upgrade Service pod.
* CHT-Sentinel pod.

## Example deployments

Below are examples from 2 Ministries of Health (MoH) and Medic hosted instances.  Users are total provisioned users, fewer users may be active in a given month.

> ### MoH Kenya

Reviewer: @mrjones-plip let's not use the name of the deployment before asking for explicit consent from the MoHes. I suggest we use `Example 1`, `Example 2` instead. Of course, this is also now in the version control of this repo, but we should definitely not use the real deployment names.

Author: gotcha - fixed in 6387fe8f

|            Cores: | 32 (4 VMs x 8 cores/ea)   |
|              RAM: | 64 GB (4 VMs x 16GB/ea)   |
|             Disk: | 600 GB (4 VMs x 150GB/ea) |
| Multi-Node Couch: | Yes                       |
|            Users: | 12k                       |

<!-- 
sources:

May 2nd,2024 Slack
https://medic.slack.com/archives/C06TP97HRMZ/p1714639214379159

-->

### Medic

Medic hosts 46 production instances on Amazon [Elastic Kubernetes Service](https://docs.aws.amazon.com/eks/latest/userguide/what-is-eks.html) (EKS).

|                   |                                                 |
|------------------:|:------------------------------------------------|
|    CHT Instances: | 46                                              |
|          Hosting: | k8s on EKS                                      |
|            Cores: | 64 (8 VMs x 8 Cores/ea)                         |
|              RAM: | 256 GB (8 VMs x 32GB/ea)                        |
|             Disk: | 8.6 TB (8 VMS x 500GB/ea + 46 CHTs x ~100GB/ea) |
| Multi-Node Couch: | 44 No, 2 Yes                                    |
|            Users: | 25.7k                                           |

> <!-- 

Reviewer: @mrjones-plip, if I'm not mistaken, this is not public data, so we shouldn't add it to our public documentation repository. Please remove it from this file until we receive explicit consent from MoHes to publish information about their deployments.

title: Kubernetes vs Docker
linkTitle: Kubernetes vs Docker
weight: 5
aliases:
  -  /hosting/vertical-vs-horizontal
---

{{< hextra/hero-subtitle >}}
  Options for installing CHT applications
{{< /hextra/hero-subtitle >}}

> Medic recommends [Docker](/hosting/4.x/docker/) for most deployments. Only use [Kubernetes](/hosting/4.x/kubernetes/) for large, multi-tenant or specific hardware constrained deployments.

Reviewer: ```suggestion
Medic recommends [Docker](/hosting/4.x/docker/) for small scale deployments. Use [Kubernetes](/hosting/4.x/kubernetes/) for large, multi-tenant or specific hardware constrained deployments.
```

Reviewer: Your sentence assumes most of CHT deployments are small scale deployments which I am unsure is accurate.

Author: My sentence actually does not assume most CHT deployments are small scale (eg only 100s of CHWs).  We have proven that Docker can support large scale deployments.  We know that all CHT deployments that are this shape can be run Docker.  Removing "only" here I think softens it enough.

fixed in 65ee12f7

Application development for both [CHT 3.x]({{< relref "hosting/3.x/app-developer" >}}) and [CHT 4.x]({{< relref "hosting/4.x/app-developer" >}}) should use Docker as well.

## Docker 

Docker's has stood the test of time and has many benefits:

* Easy to deploy by just downloading a compose file.
* Trivially scales down to a 4 core laptop and [way up to a 48 core server](https://forum.communityhealthtoolkit.org/t/investigate-adding-more-shards-as-a-potential-avenue-for-improved-performance/4831). 
* Easily integrates with existing Docker based solutions like [Watchdog](/hosting/monitoring/) or other services running in Docker.
* Backup and restore can be achieved via simple file management via [bind mounted volumes](https://docs.docker.com/engine/storage/bind-mounts/).
* Just enough complexity to securely and soundly run the CHT with out extra complexities of Kubernetes 
* Natively supports one click CHT upgrades in the admin GUI

Docker Compose deployment has the following services when running the CHT:

* A service for each CouchDB instance.
* A CHT API service.
* HAProxy Healthcheck service.
* HAProxy service.

> * Upgrade Service service.

Reviewer 2: > Upgrade Service service.

haha

Author: :sweat_smile:

* A CHT Sentinel service.
* An nginx service to act as a reverse proxy and terminate TLS connections

## Kubernetes (k8s)

{{< callout type="warning" >}}

> Only use Kubernetes on bare-metal machines with 8 cores or fewer or when dozens of CHT instances are needed in a multi-tenant deployment.

Reviewer 2: > bare-metal machines with 8 cores or fewer

I'm not sure we should stipulate the number of cores here. Maybe say low-end instead?

Reviewer: +1 on Diana's comment.

Additionally, I am reluctant of the use of the word "only". We can do recommendations according to deployment needs, but we shouldn't impose a tool or another.

Author: gotcha.  Lemme know if e7ad45ba doesn't address your feedback!

> Only use Kubernetes on bare-metal machines with 8 cores or fewer or when dozens of CHT instances are needed in a multi-tenant deployment.

Reviewer: ```suggestion
Use Kubernetes on bare-metal machines with 8 cores or fewer or when dozens of CHT instances are needed in a multi-tenant deployment.
```

Author: "only" was removed in another commit.

{{< /callout >}}

> Given the additional knowledge needed to run and complexities to deploy, CHT on k8s has been shown to result in confusion and outages. Docker is the preferred solution for most production deployments.

Reviewer: ```suggestion
Running CHT on k8s requires additional knowledge and can be more complex to deploy. Consider using Docker for small scale CHT deployments. 
```

Author: Keeping half the suggestion, see fb2b2d68

> K8s deployments can not safely use the one click upgrade button in the admin GUI.

Reviewer: Do you have a link to forum posts or data behind this statement to explain the why and what happens if people use that button? I think it will help understand what's lacking currently to have this possible.

Author: Yeah - good point!  I think this can actually be addressed (if you back up your helm charts and always use helm charts deply or restore, it works fine).  I'll just remove this for now.

fixed in 19281800

> Otherwise, k8s offers:

Reviewer: ```suggestion
k8s offers:
```
